### ISSUE TITLE
[2.17.0] GuardedBy on lambda

### ISSUE THREAD

**Issue by ben-manes** [google/error-prone#3640]
The constructor assigns a `Consumer` to which is invoked under a lock. However the checker now fails by assuming it was invoked in the constructor.

```java
final Consumer<Node<K, V>> accessPolicy;
...
accessPolicy = (evicts() || expiresAfterAccess()) ? this::onAccess : e -> {};
...
/** Drains the read buffer. */
@GuardedBy("evictionLock")
void drainReadBuffer() {
  if (!skipReadBuffer()) {
    readBuffer.drainTo(accessPolicy);
  }
}
```

```console
/Users/ben/projects/caffeine/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java:270: error: [GuardedBy] This access should be guarded by 'this.evictionLock', which is not currently held
    accessPolicy = (evicts() || expiresAfterAccess()) ? this::onAccess : e -> {};
                                                        ^
    (see https://errorprone.info/bugpattern/GuardedBy)
1 error
```

**Comment by graememorgan** [comment]
> However the checker now fails by assuming it was invoked in the constructor.

I think we need to improve the diagnostic message here, really. It's not so much failing because it assumes it's invoked in the constructor, it's failing because it can't work out whether the lock will be held when it's invoked.
